common/models/user.js

```diff
@@ -382,14 +396,114 @@ module.exports = function(User) {
     }
   );
 
+  User.requestAuthLink = function requestAuthLink(email) {
+    if (!isEmail(email)) {
+      return Promise.reject(
+        new Error('The submitted email not valid.')
+        );
+    }
+
+    var userObj = {
+      username: 'fcc' + uuid.v4().slice(0, 8),
+      email: email,
+      emailVerified: false
+    };
+    return User.findOrCreate$({ where: { email: userObj.email }}, userObj)
+      .map(([ err, user, isCreated ]) => {
+        if (err) {
+          return dedent`
+           Oops, something is not right, please try again later.`;
+        }
+
+        if (!isDev) {
+          const minutesLeft = getWaitPeriod(user.emailAuthLinkTTL);
+          if (minutesLeft) {
+            const timeToWait = minutesLeft ?
+              `${minutesLeft} minute${minutesLeft > 1 ? 's' : ''}` :
+              'a few seconds';
+            debug('request before wait time : ' + timeToWait);
+            return dedent`
+              Please wait ${timeToWait} to resend an authentication link.`;
+          }
+        }
+
+        let emailTemplate = 'user-request-sign-in.ejs';
+        if (isCreated) {
+          emailTemplate = 'user-request-sign-up.ejs';
+        }
+
+        // create a temporary access token with ttl for 1 hour
+        user.createAccessToken({ ttl: 60 * 60 * 1000 }, (err, token) => {
+          if (err) { throw err; }
+
+          const { id: loginToken } = token;
+          const loginEmail = user.email;
+          const renderAuthEmail = loopback.template(path.join(
+            __dirname,
+            '..',
+            '..',
+            'server',
+            'views',
+            'emails',
+            emailTemplate
+          ));
+          const mailOptions = {
+            type: 'email',
+            to: user.email,
+            from: 'Team@freecodecamp.com',
+            subject: 'Free Code Camp - Authentication Request!',
+            text: renderAuthEmail({
+              loginEmail,
+              loginToken
+            })
+          };
+          if (!isDev) {
+            this.email.send(mailOptions, err =>{
+              if (err) { throw err; }
+            });
+          } else {
+            console.log('~~~~\n' + mailOptions.text + '~~~~\n');
+          }
+          user.emailAuthLinkTTL = token.created;
+          user.save(err =>{ if (err) { throw err; }});
+        });
+
+        return dedent`
+           If you entered a valid email, a magic link is on its way.
+           Please follow that link to sign in.`;
+      })
+      .map((msg) => {
+        if (msg) { return msg; }
+        return dedent`
+         Oops, something is not right, please try again later.`;
+      })
+      .catch(error => {
```

Comment:
Errors must be dealt with by express here. Do not use a catch here unless modifying the error message and then rethrowing